src/main/java/javax/measure/unit/ProductUnit.java

```diff
@@ -69,17 +84,50 @@ private ProductUnit(Element[] elements) {
         _elements = elements;
     }
 
+    private static final LoadingCache<ProductUnitKey, Unit<? extends Quantity>> INSTANCE_CACHE =
+            Caffeine.newBuilder()
+                    .maximumSize(4096)
+                    .build(key -> doGetInstance(key.leftElems, key.rightElems));
+
+    private static final class ProductUnitKey {
+        private final Element[] leftElems;
+        private final Element[] rightElems;
+        private final int hashCode;
+
+        private ProductUnitKey(Element[] leftElems, Element[] rightElems) {
+            this.leftElems = leftElems;
+            this.rightElems = rightElems;
+            this.hashCode = 31* Arrays.hashCode(leftElems) + Arrays.hashCode(rightElems);
```

Comment:
[learning question] How did you arrive at this formula?

Reply:
That's a common pattern you see in other classes like String or List. I don't deviate from it unless a benchmark shows me that the hash doesn't spread well.

See https://stackoverflow.com/questions/299304/why-does-javas-hashcode-in-string-use-31-as-a-multiplier for some history behind it.

Reply:
That's a common pattern you see in other classes like String or List. I don't deviate from it unless a benchmark shows me that the hash doesn't spread well.

See https://stackoverflow.com/questions/299304/why-does-javas-hashcode-in-string-use-31-as-a-multiplier for some history behind it.